### PR TITLE
Emitter options

### PR DESCRIPTION
**Pull request by timotheeguerin**
#538

This resolve a few problem we have with configuring emitters:
- currently `compilerOptions` contains emitter specific options(`swaggerOutputFile` and `serviceCodePath`)
- some other options use `miscOptions` which is not typed and can easily conflict across emitters

Change:
- Allow the value of the emitter object in the `cadl-project.yaml` config be an object of options
  ```yaml
  emitters:
    "@cadl-lang/openapi3":
      outputFile: "./my-openapi.json"
  ```
- Using `--emit` will only specify the list of emitter to use but still use the options from the cadl-project if any.
- Emitter can export the lib via `$lib` which contains the options schema and can do the validation.
- Remove emitter specific options from `compilerOptions`

Validate for invalid options:
image

Follow up:
- Need to find a way to pass those options via the CLI as well.

### CODE REVIEW

File: packages/compiler/core/types.ts [hunk 1]
```diff
@@ -1252,18 +1253,20 @@ export interface CadlLibraryDef<
    * Provide names for emitters if there is multiple.
    */
   readonly emitter?: {
-    names?: readonly E[];
+    options?: JSONSchemaType<E>;
```
Comment (timotheeguerin) [inline packages/compiler/core/types.ts:1256]: Currently the options are a JSon schema which lets the options be a complex object with complex validation
image

Should we stick to have that and having to use `ajv` to validate the options or should make that a much simpler system which only allows `key-> value` with a simple primitive validation for the value

Comment (nguerrera) [reply packages/compiler/core/types.ts:1256]: Good question. This seems nice, but my only hesitation is if there are any consequences to pulling in this dependency on JSONSchemaType in in core types. @beterlson for thoughts.

Comment (bterlson) [reply packages/compiler/core/types.ts:1256]: How does the validation work if I have an emitter that needs some custom options?

Comment (timotheeguerin) [reply packages/compiler/core/types.ts:1256]: https://github.com/microsoft/cadl/blob/07d376074de4480ce0a58ff7b892369328e30e3d/packages/openapi3/src/lib.ts#L7-L14
This is how an emitter defines the options. All emitter options are "custom". Was that the question?

Comment (markcowl) [reply packages/compiler/core/types.ts:1256]: I am not following how this looks on the command line, does it mean we can provide json for emitter options on the command line, that we specify a configuration file for the emitter optiosn on the command line, or something else?

Comment (markcowl) [reply packages/compiler/core/types.ts:1256]: But right now we can set emitter options with `--option key=value` will this be removed when we go to a more complex emitter option schema?

File: packages/compiler/core/cli.ts [hunk 2]
```diff
@@ -54,51 +54,51 @@ async function main() {
...
             describe:
               "The output path for generated artifacts.  If it does not exist, it will be created.",
           })
-          .option("option", {
+          .options("options", {
             type: "array",
             string: true,
             describe:
-              "Key/value pairs that can be passed to Cadl components.  The format is 'key=value'.  This parameter can be used multiple times to add more options.",
+              "Key/value pairs that can be passed to Cadl components.  The format is 'emitter=value'.  This parameter can be used multiple times to add more options.",
```
Comment (markcowl) [inline packages/compiler/core/cli.ts:67]: `emitter = value`, where value looks like what?  Generally the emitter options are keyed options, so does this become:
`--options emitter=key=value`, is 'emitter' the keyword 'emitter', or is it the full emitter name?  Is it something like `emitter.key`=value`?  We need to update the documentation, and potentially the providerhub template if the cli is significantly changing this format

File: packages/compiler/core/types.ts (hunk 1 above)
Comment (markcowl) [reply packages/compiler/core/types.ts:1256]: I like this, this keeps options between emitters from clashing and makes the command very explicit.  I think, from a  usability perspective, we will want to make it possible to set emitter-specific options from the command line similar to this.

We should also look through the documentation to see where we may need to update the docs for the new command line option setting

Comment (markcowl) [reply packages/compiler/core/types.ts:1256]: But we will also want the command line help message to make this easy for users to discover, probably with an example (let's say from the OpenAPI3 emitter)

**Review by markcowl — CHANGES_REQUESTED** [review]

**Review by markcowl — COMMENTED** [review]
I really like this change, but I want to discuss options on how to roll this out - this will immediately break our templates, so we want to make sure we message this appropriately to our customers before we roll out.

**Comment by markcowl** [comment]
@timotheeguerin Here are my thoughts on ways that we could roll out the compiler options change:

- Update the PR to continue to allow the current syntax for the set of options used in the template, these are:
  - cadl-providerhub-controller
    - registrationOutputPath
  - cadl-autorest
    - azure-resource-provider-folder
    - examplesDirectory
  - cadl-azure-resource-manager
    - arm-types-path

- Update the msbuild target to check for the version of the cadl compiler, and adapt these options accordingly

Thoughts?

**Comment by markcowl** [comment]
Per offlien discussion, I am OK with this change once we have a PR to look at the impact on cadl-azure template tooling.

**Review by markcowl — APPROVED** [review]

**Pull request merged**